plugin/manager.go

```diff
@@ -0,0 +1,318 @@
+	defer m.mutex.Unlock()
+
+	user := m.db.GetUserByID(userID)
+	if user != nil {
+		return m.initializeForUser(*user)
+	}
+	return fmt.Errorf("user with id %d not found", userID)
+}
+
+func (m *Manager) initializeForUser(user model.User) error {
+
+	userCtx := compat.UserContext{
+		ID:    user.ID,
+		Name:  user.Name,
+		Admin: user.Admin,
+	}
+
+	for _, p := range m.plugins {
+		if err := m.initializeSingleUserPlugin(userCtx, p); err != nil {
+			return err
+		}
+	}
+
+	for _, app := range m.db.GetApplicationsByUser(user.ID) {
+		if conf := m.db.GetPluginConfByApplicationID(app.ID); conf != nil {
+			_, compatExist := m.plugins[conf.ModulePath]
+			app.Internal = compatExist
+		} else {
+			app.Internal = false
+		}
+		m.db.UpdateApplication(app)
+	}
+
+	return nil
+}
+
+func (m *Manager) initializeSingleUserPlugin(userCtx compat.UserContext, p compat.Plugin) error {
+	info := p.PluginInfo()
+	instance := p.NewPluginInstance(userCtx)
+	userID := userCtx.ID
+
+	pluginConf := m.db.GetPluginConfByUserAndPath(userID, info.ModulePath)
+
+	if pluginConf == nil {
+		var err error
+		pluginConf, err = m.createPluginConf(instance, info, userID)
+		if err != nil {
+			return err
+		}
+	}
+
+	m.instances[pluginConf.ID] = instance
+
+	if compat.HasSupport(instance, compat.Messenger) {
+		instance.SetMessageHandler(redirectToChannel{
+			ApplicationID: pluginConf.ApplicationID,
+			UserID:        pluginConf.UserID,
+			Messages:      m.messages})
+	}
+	if compat.HasSupport(instance, compat.Storager) {
+		instance.SetStorageHandler(dbStorageHandler{pluginConf.ID, m.db})
+	}
+	if compat.HasSupport(instance, compat.Configurer) {
+		if len(pluginConf.Config) == 0 {
+			// The Configurer is newly implemented
+			// Use the default config
+			pluginConf.Config, _ = yaml.Marshal(instance.DefaultConfig())
+			m.db.UpdatePluginConf(pluginConf)
+		}
+		c := instance.EmptyConfig()
+		yaml.Unmarshal(pluginConf.Config, c)
+		instance.ValidateAndSetConfig(c)
+	}
+	if compat.HasSupport(instance, compat.Webhooker) {
+		instance.RegisterWebhook(*m.mux.Group(pluginConf.Token))
+	}
+	if pluginConf.Enabled {
+		err := instance.Enable()
+		if err != nil {
+			// Single user plugin cannot be enabled
+			// Don't panic, disable for now and wait for user to update config
+			log.Printf("Plugin initialize failed for user %s: %s. Disabling now...", userCtx.Name, err.Error())
+			pluginConf.Enabled = false
+			m.db.UpdatePluginConf(pluginConf)
+		}
+	}
+	return nil
+}
+
+func (m *Manager) createPluginConf(instance compat.PluginInstance, info compat.Info, userID uint) (*model.PluginConf, error) {
+	pluginConf := &model.PluginConf{
+		UserID:     userID,
+		ModulePath: info.ModulePath,
+		Token:      auth.GeneratePluginToken(),
+	}
+	if compat.HasSupport(instance, compat.Configurer) {
+		pluginConf.Config, _ = yaml.Marshal(instance.DefaultConfig())
+	}
+	if compat.HasSupport(instance, compat.Messenger) {
+		app := &model.Application{
+			Token:       auth.GenerateApplicationToken(),
```

Comment:
I would say it would be more "human-friendly" but it does not bring any real benefit both to users and developers, and requires extra code both on the server and client side. Let's leave this for later.

Reply:
Oke.